### PR TITLE
Add WCF update extension

### PR DESCRIPTION
**Pull request by SimonaLiao**
**Add WCF update extension**
Adds a WCFUpdater extension which will automatically port WCF projects to use CoreWCF services. PR also includes unit tests for all classes and an integration test with the BeanTrader example. 

**PR Description**
Here are most files in my extension library:
1. SourceCodeUpdater: updates source code file such as program.cs. Deals with using directives, ServiceHost instance, and registering configuration. (Currently supports serviceMetadata and serviceDebug behaviors.)
2. ConfigUpdater: analyzes the config file and updates it to create a new wcf.config
3. PackageUpdater: updates the project file by changing sdk & package references
4. UpdaterFactory: constructs the above updaters
5. UpdateRunner: run and write all the changes

Addresses #1083

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/upgrade-assistant?pullRequest=1237) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/upgrade-assistant?pullRequest=1237) before we can accept your contribution.<br/><br/>:x: SimonaLiao [sign now](https://cla.dotnetfoundation.org/dotnet/upgrade-assistant?pullRequest=1237)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/upgrade-assistant?pullRequest=1237) it.</sub>

**Comment by brandonh-msft** [comment]
Thanks @SimonaLiao ! Can you do a round of fixes [for the code/style analyzers](https://dev.azure.com/dnceng/public/_build/results?buildId=1912783&view=logs&j=396e4757-3dc0-597a-e469-5feb3123c7be&t=44c646e6-db1b-5521-5b76-c5c2c57c3a93&l=135)?

Don't worry about the arcade build issue on the `xunit` references for now - I'll have a look into that as it doesn't appear to be related to your change.

**Review by brandonh-msft — CHANGES_REQUESTED** [review]
Thanks @SimonaLiao ! Can you do a round of fixes [for the code/style analyzers](https://dev.azure.com/dnceng/public/_build/results?buildId=1912783&view=logs&j=396e4757-3dc0-597a-e469-5feb3123c7be&t=44c646e6-db1b-5521-5b76-c5c2c57c3a93&l=135)?

Don't worry about the arcade build issue on the `xunit` references for now - I'll have a look into that as it doesn't appear to be related to your change.

File: tests/tool/Integration.Tests/E2ETest.cs
Comment (brandonh-msft) [inline tests/tool/Integration.Tests/E2ETest.cs:44]: This is causing a compilation error in the build
```suggestion
```

**Review by brandonh-msft — CHANGES_REQUESTED** [review]

**Review by brandonh-msft — CHANGES_REQUESTED** [review]
try-convert needs to stay removed from the repo; this change was merged to `main` already.

**Review by brandonh-msft — CHANGES_REQUESTED** [review]
Blocking review. Once one of the WCF folks signs off, I'll approve (assuming builds are all good)

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateChecker.cs
Comment (brandonh-msft) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateChecker.cs:94]: prefer `is not null`
```suggestion
            if (doc.Root.Attribute("Sdk") is not null && !doc.Root.Attribute("Sdk").Value.Contains("Web"))
```

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs
Comment (brandonh-msft) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs:116]:
```suggestion
                Logger.LogInformation("This project is not applicable for updating to CoreWCF since references to WCF were not found. No more work needs to be done and this step is complete.");
```

Comment (brandonh-msft) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs:116]:
```suggestion
                var msg = "This project is not applicable for updating to CoreWCF since references to WCF were not found. No more work needs to be done and this step is complete.";
                Logger.LogInformation(msg);
```

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs
Comment (brandonh-msft) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs:117]:
```suggestion
                return Task.FromResult(new UpgradeStepInitializeResult(UpgradeStepStatus.Complete, msg, BuildBreakRisk.Low));
```

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs
Comment (brandonh-msft) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs:190]:
```suggestion
                Logger.LogWarning("Can not find .cs file with Main() method. The project is not applicable for automated WCF update. No more work needs to be done and this step is complete.");
```

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs
Comment (brandonh-msft) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs:194]:
```suggestion
                Logger.LogWarning("Found more than one .cs file with Main() method. The project is not applicable for automated WCF update. No more work needs to be done and this step is complete.");
```

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs
Comment (brandonh-msft) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs:204]:
```suggestion
                Logger.LogWarning("Did not find .config file that configures system.serviceModel. The project is not applicable for automated WCF update. No more work needs to be done and this step is complete.");
```

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdateRunner.cs
Comment (mjrousos) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdateRunner.cs:90]: As mentioned above, I think these calls (RemoveOldCode and AddTemplateCode - and, really, even UpdateDirectives) are likely to always happen together so it may make sense to have a single source code updater method that wraps them all.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdaterFactory.cs
Comment (mjrousos) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdaterFactory.cs:54]: Do we expect any exceptions from the SourceCodeUpdater? I wouldn't expect that component to throw so we probably don't need the try/catch here.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdaterFactory.cs
Comment (mjrousos) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdaterFactory.cs:39]: As in other places, if we can catch a more specific exception, that would be good. It will avoid accidentally catching some unexpected catastrophic event that really ought to not be handled.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdaterFactory.cs
Comment (mjrousos) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdaterFactory.cs:61]: The name for this method is confusing. It's getting source code updaters but it's called GetDirectiveUpdaters.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdateRunner.cs
Comment (mjrousos) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdateRunner.cs:160]: Rather than use a Dictionary for this, could we have a simple type for storing WcfUpgradeContext? It would be more reliable (if something changes in that model type, things won't build until all users agree on the shape of it unlike a Dictionary which would happily build and then fail at run time).

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs
Comment (mjrousos) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs:46]: Does the ConfigUpdater belong here? I'd have thought you'd want to run before it in order to remove the system.ServiceModel section.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs
Comment (mjrousos) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs:53]: SourceUpdaterStep also (and maybe ConfigUpdaterStep?) to ensure you run before them?

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs
Comment (mjrousos) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs:122]: Technically, ILogger's are usually specific to a type (so that it's possible to correlate traces to the types that emitted them) so you shouldn't share this class's ILogger. Instead, you can get an ILoggerFactory from DI and use that to create loggers for other classes, as neede.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs
Comment (mjrousos) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs:106]: A class with fields for expected paths might be better than a dictionary. More strongly typed.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs
Comment (mjrousos) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs:141]: The BuildBreakRisk is probably still medium here. That means that if the user proceeds with applying the change there's a "medium" chance that they'll need to go back and make manual updates afterwards to get the project building again.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs
Comment (mjrousos) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs:172]: Since the project will have changed directly on disk and not via the project interfaces, it's a best practice to reload the project workspace. That way, we can be sure that future upgrade steps will see your changes reflected in what's loaded. You can do that with `IUpgradeContext.ReloadWorkspaceAsync`.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs
Comment (mjrousos) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs:164]: Rather than using `_path!` it would be better to check that `_path` is not null. It shouldn't ever happen, but who knows if a bug might creep in allowing Apply to be erroneously called without Initialize. If _path were null, you could just return a failing result. And if it's not, then you won't need !.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateChecker.cs
Comment (mjrousos) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateChecker.cs:86]: The IProject interface (retrieved from IUpgradeContext.CurrentProject can do these queries more directly. Either way is fine, though.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdaterFactory.cs
Comment (mjrousos) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdaterFactory.cs:16]: Rather than having static methods, this class could be made available in dependency injeciton (similar to the upgrade step) and other classes could request it in their constructors. That would also allow this class to get an instance of ILoggerFactory in its ctor which could be used to create the loggers for all the various updaters.

**Review by mjrousos — DISMISSED** [review]
Looks good! Thanks for putting this together, @SimonaLiao! I've put some comments and suggestions in-line but nothing blocking.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.csproj
Comment (SimonaLiao) [reply src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.csproj:4]: I switched from 2.0 to 2.1 because 2.1 has the overload option Contains(string, StringComparison), which could save a little bit time compared to Contains(string). But I think the flexibility might be more preferable in this case?

@mconnew Matt, What's your opinion on this? :)

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdaterFactory.cs
Comment (SimonaLiao) [reply src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdaterFactory.cs:39]: I added a lot of try/catch for the UpdateRunner + UpdateFactory and was just generally concerned about unexpected errors happening and I thought it would be better to catch it as early as possible instead of letting it bubble up? I am wondering what would be the best practice for the error handling in this case?

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.csproj
Comment (mjrousos) [reply src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.csproj:4]: Yea, the new `String.Contains` overload in 2.1 is nice. But you can get the same behavior on netstandard2.0 by using `string.IndexOf(string, StringComparison)` and just checking that the result is >= 0.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdaterFactory.cs
Comment (SimonaLiao) [reply src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdaterFactory.cs:61]: It's a different type of SourceCodeUpdater! It's for .cs files that only needs to add the using directives for coreWCF, such as BeanTrader.cs. I could change it to a different name to be more clear?

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs
Comment (SimonaLiao) [reply src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs]: Hi Mike! Sometimes after the WCF update, there might be some compile errors left in the source code that needs manual updates. Will those errors break the default source code updater? If not I can add WCF update step before them! :)

Comment (mjrousos) [reply src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs]: SourceCodeUpdater doesn't need a clean build, so it's ok to move this one first.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/ConfigUpdater.cs
Comment (mconnew) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/ConfigUpdater.cs:78]: It's not clear what the return type is here. Either rename the method to be clearer what's returned, or explicitly specify the data type for uri. Maybe GetSchemeToBasepathMapping if renaming?

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/SourceCodeUpdater.cs
Comment (mconnew) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/SourceCodeUpdater.cs:70]: I think d.Trim() might be better. There are a couple of potential issues with this. The NewLine character for the currently running OS might be different than what's in the file. Linux and Windows have a different NewLine sequence. You could have a combination which leaves a character behind which would mean it wouldn't match the line on the Contains call a few lines below. Also some people have extra blank lines at the end of the line before the new line and the current code would break on that. I belied string.Trim() will cover all those cases.

File: tests/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.Test/ConfigUpdaterTest.cs
Comment (mconnew) [inline tests/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.Test/ConfigUpdaterTest.cs:123]: Use string.Empty instead of ""

Comment (SimonaLiao) [reply tests/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.Test/ConfigUpdaterTest.cs:123]: The attribute argument here has to be a constant expression and it's not letting me use string.Empty

Comment (brandonh-msft) [reply tests/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.Test/ConfigUpdaterTest.cs:123]: Yup that's correct - can't use `string.Empty` here.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.csproj
Comment (mconnew) [reply src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.csproj:4]: Isn't the only difference is supported frameworks between netstandard2.0 and netstandard2.1 whether you can run on .NET Framework or not? Unless I've missed something, every other in support framework can use 2.1

Comment (mjrousos) [reply src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.csproj:4]: Yes, but it's possible that in the future we may want to re-use some of this code in a VS extension to enable upgrade scenarios to work there and we'd need NetFx if we wanted to run in-proc.

**Comment by brandonh-msft** [comment]
@SimonaLiao if you have addressed any asks/comments w/in this PR, please use the `Resolve Conversation` button so we know what's been done and what hasn't - this will expedite re-reviews :)

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdaterFactory.cs
Comment (SimonaLiao) [reply src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdaterFactory.cs]: The using directive updaters are constructed by a different constructor in the` SourceCodeUpdater` class and they can only be used to update directives, and can't do template code/remove old code stuff because in their constructor no template is passed in!

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/PackageUpdater.cs
Comment (SimonaLiao) [reply src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/PackageUpdater.cs]: The IProjectFile would be really convenient for managing packages! But it would complicate the requirements to construct a  `PackageUpdater`. Now the constructor only requires a XDocument and it's easy to test. I am using the `GetFile()` method in `WCFUpdater.cs` to get easy access to project file and for the updaters it might be better to have some separation.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.csproj
Comment (SimonaLiao) [reply src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.csproj:4]: Hi Mike! Thanks for providing those contexts and that makes sense. I will change it back to .NET2.0.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdateRunner.cs
Comment (SimonaLiao) [reply src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdateRunner.cs:99]: Hi Mike! That's a great suggestion. And I will implement it in my next PR (one with all extra features I implemented) to avoid repeated merge conflicts. I have summarized those improvements I will make in this issue: https://github.com/dotnet/upgrade-assistant/issues/1270

**Review by HongGit — APPROVED** [review]

**Review by brandonh-msft — APPROVED** [review]
🚢🎉

**Pull request merged**
